pcdscalc/pmps.py

```diff
@@ -0,0 +1,256 @@
+    1.82E3,
+    2.0E3,
+    2.3E3,
+    ]
+
+
+def select_bitmask_boundaries(line):
+    """
+    Given a line, select the bitmask boundaries to use.
+
+    These are hard-coded for now but in theory this could be loaded from
+    a database, an EPICS PV, or some other source.
+
+    Parameters
+    ----------
+    line: str
+        String representation of which line's bitmask to use.
+        If the string begins with "l" or "h" (lfe, hxr), we'll
+        use the hard-xray bitmask.
+        If the string begins with "k" or "s" (kfe, sxr), we'll
+        use the soft-xray bitmask.
+
+    Returns
+    -------
+    boundaries: list of floats
+    """
+    if line.lower()[0] in ['l', 'h']:
+        return LFE
+    if line.lower()[0] in ['k', 's']:
+        return KFE
+    raise ValueError(f'{line} is neither lfe or kfe!.')
+
+
+def get_bitmask(lower, upper, allow, line, bounds=None):
+    """
+    Given a range of eV values, calculate the appropriate pmps bitmask.
+
+    This saves you the effort of checking up on the eV ranges and
+    remembering how the bitmask is assembled.
+
+    If the lower or upper fall within a range (not on the border), then that
+    range is always considered unsafe.
+
+    The rules for the bitmasks are:
+        - The nth bit of the bitmask represents a range from the n-1th value
+          to the nth value
+        - If the bit is 1, the range is allowed and is safe.
+        - If the bit is 0, the range is not allowed and is not safe.
+        - eVs above the highest boundary are never safe
+        - negative eVs are never safe
+
+    Some examples:
+        - bitmask 00 -> no beam range is safe
+        - bitmask 01 -> only the lowest boundary and below are safe
+        - bitmask 11 -> only the two lowest boundaries and below are safe
+        - bitmask 10 -> only the space between the lowest and second-lowest
+                        boundaries are safe.
+        - bitmask all 1s -> every eV is OK except for above the highest
+                            boundary and except for negative eV
+
+    Parameters
+    ----------
+    lower: number
+        The value in eV for the lower bound of the range.
+
+    upper: number
+        The value in eV for the upper bound of the range.
+
+    allow: bool
+        True if we want a bitmask that only includes this range,
+        False if we want a bitmask that only excludes this range.
+
+    line: str
+        String representation of which line's bitmask to use.
+        If the string begins with "l" or "h" (lfe, hxr), we'll
+        use the hard-xray bitmask.
+        If the string begins with "k" or "s" (kfe, sxr), we'll
+        use the soft-xray bitmask.
+
+    bounds: list of numbers, optional
+        Custom boundaries to use instead of the default soft-xray
+        or hard-xray lines. Useful for testing.
+
+    Returns
+    -------
+    bitmask: int
+    """
+    # Be lenient on input args
+    if upper < lower:
+        lower, upper = upper, lower
```

Comment:
But maybe not too lenient;
`assert lower >= 0 and upper >= 0`?

Reply:
As a warning to a wayward user who puts in a negative, because it's probably in error? I was being lenient and just accepting it and calculating appropriately, e.g. `lower, upper = -10, -1` would give an all zero bitmask.

Reply:
Yeah, I think mainly for users passing in calculated ranges and messing up somehow. Doesn't look like it'd really affect the logic, of course.